straxen/plugins/event_processing.py

```diff
@@ -733,58 +734,153 @@ def cs2_to_e(self, x):
 
 
 @export
+@strax.takes_config(
+    strax.Option(name='pre_s1_area_threshold', default=1e3,
+                 help='Only take S1s larger than this into account '
+                      'when calculating PeakShadow [PE]'),
+    strax.Option(name='pre_s2_area_threshold', default=1e4,
+                 help='Only take S2s larger than this into account '
+                      'when calculating PeakShadow [PE]'),
+    strax.Option(name='deltatime_exponent', default=-1.0,
+                 help='The exponent of delta t when calculating shadow'),
+    strax.Option(name='time_window_backward', default=int(1e9),
+                 help='Search for peaks casting shadow in this time window [ns]')
+)
 class EventShadow(strax.Plugin):
     """
-    This plugin can calculate shadow at event level.
-    It depends on peak-level shadow.
-    The event-level shadow is its first S2 peak's shadow.
-    If no S2 peaks, the event shadow will be nan.
-    It also gives the position infomation of the previous S2s
-    and main peaks' shadow.
+    This plugin can calculate shadow for main S1 and the 1st S2 peak in events.
+    The event's shadow is its 1st S2 peak's shadow.
+    It also gives the position information of the primary peaks.
+    References:
+        * v0.1.1 reference: xenon:xenonnt:ac:prediction:shadow_ambience
     """
-    __version__ = '0.0.8'
+    __version__ = '0.1.1'
     depends_on = ('event_basics', 'peak_basics', 'peak_shadow')
     provides = 'event_shadow'
+    data_kind = 'events'
+    save_when = strax.SaveWhen.EXPLICIT
+
+    def infer_dtype(self):
+        dtype = []
+        for main_peak, main_peak_desc in zip(['s1_', ''], ['main s1', 'the 1st S2 peak']):
+            for key in ['s1_time', 's2_time', 's2_position']:
+                type_str = key.split('_')[0]
+                tp_desc = key.split('_')[1]
+                dtype.append(((f'{tp_desc} shadow casting from {type_str} to {main_peak_desc} [PE/ns]', f'{main_peak}shadow_{key}'), np.float32))
+                dtype.append(((f'area of primary {type_str} casting {tp_desc} shadow on {main_peak_desc} [PE]', f'{main_peak}pre_area_{key}'), np.float32))
+                dtype.append(((f'time difference from the primary {type_str} casting {tp_desc} shadow to {main_peak_desc} [ns]', f'{main_peak}shadow_dt_{key}'), np.int64))
+                if 'time' in key:
+                    dtype.append(((f'Time difference from the nearest primary {type_str} to {main_peak_desc} [ns]', f'{main_peak}nearest_dt_{type_str}'), np.int64))
+        for x in ['x', 'y']:
+            dtype.append(((f'{x} of primary s2 peak casting time shadow on the 1st S2 peak [cm]', f'pre_{x}_s2_time'), np.float32))
```

Comment:
What is the difference between primary S2, main S2 and 1st S2?

Reply:
For example, I don't understand yet the description of following dtype:

- "time shadow casting from s1 to main s1 [PE/ns]"  (btw, what is time shadow?)
- "area of primary s1 casting time shadow on main s1 [PE]"
-

Reply:
@skazama I updated the dtype descriptions. Hope you can get the points now. :)